### ISSUE TITLE
[Awaiting Payment 2024-01-30] [$2000] We can send an empty message using dictation on iPhone

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#18121]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open App
2. Go to any chat
3. Press on the message composer
4. Press the mic icon from the keyboard and do not speak anything 
5. Press the mic icon again to stop dictation

## Expected Result:
If we have not spoken anything then send button should have to be disabled

## Actual Result:
Send button is enabled on an empty message and we can send the empty message in the chat

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.7
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/235032144-b663de0d-7263-43b0-b0e7-7af7696c9157.MP4

https://user-images.githubusercontent.com/43996225/235032228-0752377a-c7ce-41ea-b54d-4bdbd11bf5d0.MP4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @jayeshmangwani 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1682627289846559

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01fe24292fbc44e5d8</li>
        <li>Upwork Job ID: 1651956178824339456</li>
        <li>Last Price Increase: 2023-05-08</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @trjExpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by hungvu193** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
We can send an empty message using dictation on iPhone
### What is the root cause of that problem?
We are checking if comment is empty to disable the send button, however the result return from empty dictation in iPhone was a escape character `\ufffc`, that's why our regex didn't work and button still can be enable.
https://github.com/Expensify/App/blob/6fb096948d862ee008445f809cff3ecd51b36f55/src/pages/home/report/ReportActionCompose.js#L578
### What changes do you think we should make in order to solve the problem?
Add more condition to our isCommentEmpty's regex to prevent such thing. So our regex will look like:
```javascript
isCommentEmpty: !!newComment.match(/^(\s|\ufffc)*$/),
```

### What alternative solutions did you explore? (Optional)
None.

**Comment by trjExpensify** [comment]
I can reliably reproduce this. Moving on `External`.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01fe24292fbc44e5d8

**Comment by MelvinBot** [comment]
Current assignee @trjExpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @Santhosh-Sellavel (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @flodnv (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by DmytroDumas** [comment]
I think user don't speak anything, but because of surrounding situation, there are some problems.

**Comment by MelvinBot** [comment]
📣 @DmytroDumas! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/231263719-a8a3b7d3-36a0-4d7f-8ad4-cebe1f560fa4.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by hellohublot** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

We can send an empty message using dictation on iPhone

### What is the root cause of that problem?

When we use voice input, the system don't care whether the content is recognized or not, the system always add a `\uFFFC` at the end of the text

### What changes do you think we should make in order to solve the problem?

The system always automatically add a `\uFFFC`, so we need to delete it one more time when we hit the `delete button`. because it always ends up with one more character.

When I tested the Slack app, I found that they replaced all the `\uFFFC`.

Moreover, sending this character to other platforms may display `OBJ` or others
So we should delete all the `\uFFFC`

```javascript
onChangeText={comment => this.updateComment(comment.replace(/\uFFFC/g, ''), true)}
```

**Comment by svendev2024** [comment]
React Native is a powerful tool for developing cross-platform mobile apps, and I have extensive experience in working with it. I have encountered similar issues in the past and am confident that I have the skills and expertise to resolve this for you.

To fix the issue, we need to disable the send button if the user has not spoken any message. We can achieve this by using the onStartListening and onEndListening methods provided by the dictation library. These methods allow us to track the state of the dictation session, and we can use this information to enable/disable the send button accordingly.

Also we can fix as follow code:
onChangeText={comment => this.updateComment(comment.replace(/\uFFFC/g, ''), true)}

Contributor details
Your Expensify account email: sunshine950101@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01fe1403955a8aad55

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by Santhosh-Sellavel** [comment]
Quick thoughts

Seems the @hellohublot's proposal is straight forward one [here](https://github.com/Expensify/App/issues/18121#issuecomment-1527754639)

We should address this at all places, not just here
> We can even do all the replacements in `BaseTextInput`, because they should not be sent to other platforms

cc: @flodnv

**Comment by hellohublot** [comment]
@flodnv @Santhosh-Sellavel Thanks. 
I investigated further and it is indeed a React Native issue.
I updated my [proposal](https://github.com/Expensify/App/issues/18121#issuecomment-1527754639) and the root cause, I also submitted a [PR](https://github.com/facebook/react-native/pull/37188) to the upstream. and I would prefer Solution A

**Comment by Santhosh-Sellavel** [comment]
@hellohublot I'm not sure 

/\uFFFC+/ Having this check on multiple places is bad idea, we need handle this at single place i.e for every input

**Comment by hellohublot** [comment]
@Santhosh-Sellavel Thanks

Thanks, but the following video looks like we don’t need to change the `disabled state`. When every button is clicked, the system will `interrupt` the dictation first, then update the text, and then respond to the button click, so we only need to wait for the [Upstream PR](https://github.com/facebook/react-native/pull/37188) to be merged, and then update version or merge it into `Expensify/react-native`


https://user-images.githubusercontent.com/20135674/236119627-944084bc-4f04-4ea2-82ba-55aa469837b9.mp4

**Comment by hellohublot** [comment]
@flodnv It seems that another [issue](https://github.com/Expensify/App/issues/17174) will be solved too, both are iOS16+ to reproduce, I tested both can be solved

Would you mind holding the other one (it doesn't have a proposal yet), or raising the bounty on this one, because it did cost me some time and I didn't find any solution through google

**Comment by Santhosh-Sellavel** [comment]
> @flodnv It seems that another [issue](https://github.com/Expensify/App/issues/17174) will be solved too, both are iOS16+ to reproduce, I tested both can be solved
> 
> Would you mind holding the other one (it doesn't have a proposal yet), or raising the bounty on this one, because it did cost me some time and I didn't find any solution through google

Left a comment here -> https://github.com/Expensify/App/issues/17174#issuecomment-1537480278

Seems we are waiting PR review on Upstream

**Comment by trjExpensify** [comment]
We've put the other on hold to consolidate to this issue. How's it going with the PR upstream? @flodnv shouldn't we be following [this SO](https://stackoverflowteams.com/c/expensify/questions/12761) (internal ref) and also submitting a PR against our fork in the meantime while we wait?

**Comment by trjExpensify** [comment]
For `1`: https://github.com/Expensify/react-native/pulls is our fork, no?

**Comment by flodnv** [comment]
Wow, yes, no idea why my GH searches did not pull it up 🤦 Let's see how that upstream PR is going, so far it seems to be going well @hellohublot ?

**Comment by hellohublot** [comment]
@trjExpensify  @flodnv Thanks, Yes, the upstream pr has been imported to internal https://github.com/facebook/react-native/pull/37188#issuecomment-1534415302, so we are still waiting for the merge, we can wait for a while, of course, I also created a fork pr https://github.com/Expensify/react-native/pull/53, if the upstream takes too long, we also can merge the fork first

**Comment by flodnv** [comment]
Great, thanks! Let's see how this week goes.

**Comment by trjExpensify** [comment]
Works for me!

**Comment by Santhosh-Sellavel** [comment]
@hellohublot What's the status here can you bump the reviewer?

**Comment by hellohublot** [comment]
@Santhosh-Sellavel Thank you. I just asked and I'm waiting for a reply

**Comment by flodnv** [comment]
In a couple days, let's just merge the change in our fork.

**Comment by flodnv** [comment]
Movement has been happening in the external PR, thanks for your continued work @hellohublot 👍

**Comment by Santhosh-Sellavel** [comment]
Not overdue @hellohublot can you bump reviewers again so they follow, thanks!

**Comment by hellohublot** [comment]
Of course, request again

**Comment by hellohublot** [comment]
Thank you, the upstream pr has been merged. I'm not sure if fork pr is useful now, but I still cheer-pick and updated fork pr. I don't know what the next step is. Maybe we can just wait for Expensify/react-native to integrate the upstream pr and update the rc version.

**Comment by flodnv** [comment]
Thanks @hellohublot ! 

The next steps are as follows:
1. Merge our fork PR
2. Update E/App to use the updated version of the fork with the PR integrated upstream once a new version is released, and track that in another GH that I will create
3. @hellohublot you can then create a PR in this repo to update the dependency and fix this bug 

How's that sound?

**Comment by trjExpensify** [comment]
Nice, great job here!

**Comment by flodnv** [comment]
I created a task to look into updating our fork if it does not get done by someone else: https://github.com/Expensify/App/issues/19630

Unsure how long checks on [this PR](https://github.com/Expensify/react-native/pull/53) are going to take, I set myself a reminder to merge it

**Comment by trjExpensify** [comment]
We need to add the `reviewing` label here, right?

**Comment by flodnv** [comment]
Kind of not yet @trjExpensify, we're still missing the App PR. 

@hellohublot a GH action is failing for some reason https://github.com/Expensify/react-native/pull/53#issuecomment-1567296958 😕

**Comment by hellohublot** [comment]
@flodnv Thanks, already gave some thoughts about CI failures https://github.com/Expensify/react-native/pull/53#issuecomment-1567629056

**Comment by flodnv** [comment]
I've asked Rory if someone else can take over releasing our fork so we can finally use the fix... https://expensify.slack.com/archives/C01GTK53T8Q/p1686087279547249?thread_ts=1685557998.465289&cid=C01GTK53T8Q

**Comment by trjExpensify** [comment]
Nice, everybody loves to fork. 😄

**Comment by trjExpensify** [comment]
@flodnv what are we doing here?

**Comment by AndrewGable** [comment]
@hellohublot @flodnv Was https://github.com/facebook/react-native/pull/37188 CP-ed into a deploy for RN? Or which release is it included in?

**Comment by flodnv** [comment]
[Apparently](https://github.com/facebook/react-native/pull/37188#issuecomment-1559557977) it was merged here: https://github.com/facebook/react-native/commit/e8b4bb0684eed2acdb4648205893ee372f7ac1c7

**Comment by AndrewGable** [comment]
Agree it was merged, but I don't _think_ it's currently in any releases, so we might have to wait until the next react native release (unless we ask them to cherry pick it)

**Comment by flodnv** [comment]
How do you check if it's currently in a release? There would be an indicator somewhere on this link https://github.com/facebook/react-native/commit/e8b4bb0684eed2acdb4648205893ee372f7ac1c7? 

It's been "merged" (where to? 😕 ) 3 weeks ago, how often do they release?

**Comment by flodnv** [comment]
@AndrewGable do you have any guesstimate as to the ETA of that RN release?

**Comment by flodnv** [comment]
Ok so at this point we just check back in 1 month, right...?

**Comment by AndrewGable** [comment]
Yes, _unless_ we want to ask for this to be cherry picked if it fits this description: https://reactnative.dev/contributing/release-faq#what-is-a-qualified-pick-request

**Comment by trjExpensify** [comment]
Can someone remind me why we aren't merging this in our fork?

**Comment by Santhosh-Sellavel** [comment]
> Ah, I believe we are now waiting on @AndrewGable to implement [this proposal](https://expensify.slack.com/archives/C01GTK53T8Q/p1686159512184629).

@trjExpensify

**Comment by trjExpensify** [comment]
Oh, welp. So now we just sit on our hands for a bit then? Where exactly are we looking? Also, should this be downgraded to `monthly` then if we're waiting a month or two for a RN release?

**Comment by flodnv** [comment]
AFAIK we are looking here: https://github.com/facebook/react-native/releases

Specifically in the release notes @hellohublot @Santhosh-Sellavel @AndrewGable ? https://github.com/facebook/react-native/blob/main/CHANGELOG.md -- I don't see `dictation` mentioned, so maybe it'll be in the next one? Or maybe I am holding this wrong?

**Comment by Santhosh-Sellavel** [comment]
It's not tagged for release yet!

**Comment by trjExpensify** [comment]
>It's not tagged for release yet!

@Santhosh-Sellavel does the PR get a label or something?

**Comment by trjExpensify** [comment]
Has there been a change here? Still curious about that label!

**Comment by Santhosh-Sellavel** [comment]
@hellohublot Can you follow up on this and update?

**Comment by flodnv** [comment]
@AndrewGable how can we find out when they'll include it in a release? 😕

**Comment by AndrewGable** [comment]
I think it will be included in the 0.73 release which they are starting to prepare now (I believe)

**Comment by trjExpensify** [comment]
Nice, that's great news!

**Comment by trjExpensify** [comment]
Still waiting for 0.73

**Comment by trjExpensify** [comment]
Still waiting for 0.73

**Comment by flodnv** [comment]
Some 0.73 release info here: https://github.com/Expensify/App/issues/14676#issuecomment-1714793121

**Comment by trjExpensify** [comment]
Still a little bit out on that timeline!

**Comment by flodnv** [comment]
I think we're still waiting...

**Comment by flodnv** [comment]
0.73 upgrade is in progress: https://github.com/Expensify/App/pull/31558

**Comment by flodnv** [comment]
https://github.com/Expensify/App/pull/31558 is on prod!!! 😱 Can someone retest this on iPhone please? 🙏

**Comment by Santhosh-Sellavel** [comment]
@flodnv It's fixed now, tested now.

#### 1.4.24 Issue Exists
https://github.com/Expensify/App/assets/85645967/83b2f19a-5cd7-4b31-bdfb-656d93770351

#### 1.4.28 Works Fine
https://github.com/Expensify/App/assets/85645967/712f6d2b-c887-4044-8f57-8bced6310c1a

**Comment by flodnv** [comment]
Hooray!!!! 😍 

@trjExpensify finally we can issue payments here!!! 💸 

Once again, thanks for fixing this @hellohublot 👍

**Comment by trjExpensify** [comment]
Niceeee! Okay, so confirming these are the payments? 

- @hellohublot $2000 for the fix 
- @Santhosh-Sellavel $2000 for C+ reviews
- @jayeshmangwani $250 for the bug report (grandfathered rate at the time of issue creation)

**Comment by trjExpensify** [comment]
Thanks for confirming, Flo. 

@Santhosh-Sellavel - go ahead and request. 
@hellohublot - offer sent 
@jayeshmangwani - offer sent

Putting this on `Daily` and `Awaiting payment` to float it higher in my priority list :)

**Comment by trjExpensify** [comment]
@hellohublot - paid!

**Comment by trjExpensify** [comment]
@jayeshmangwani - paid! 

@Santhosh-Sellavel once you confirm you've requested, I'll go ahead and close this.

**Comment by trjExpensify** [comment]
Bump @Santhosh-Sellavel! Confirm you've requested, please.

**Comment by trjExpensify** [comment]
Mic check 1, 2.. @Santhosh-Sellavel you there? 😅

**Comment by Santhosh-Sellavel** [comment]
Will request this week. Feel free to close

**Comment by trjExpensify** [comment]
Thanks!

**Comment by Santhosh-Sellavel** [comment]
Requested on ND

**Comment by JmillsExpensify** [comment]
$2,000 approved for @Santhosh-Sellavel based on the summary above.
